Lib/ufoLib/filesystem.py

```diff
@@ -0,0 +1,523 @@
+		path2 = self._fsRootPath(path2)
+		if self.isDirectory(path1):
+			self._fs.movedir(path1, path2, create=True)
+		else:
+			self._fs.move(path1, path2)
+
+	def _fsExists(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.exists(path)
+
+	def _fsIsDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.isdir(path)
+
+	def _fsListDirectory(self, path):
+		path = self._fsRootPath(path)
+		return self._fs.listdir(path)
+
+	def _fsGetFileModificationTime(self, path):
+		path = self._fsRootPath(path)
+		info = self._fs.getinfo(path, namespaces=["details"])
+		return info.modified
+
+	# -----------------
+	# Path Manipulation
+	# -----------------
+
+	def joinPath(self, *parts):
+		if haveFS:
+			return fs.path.join(*parts)
+		else:
+			return os.path.join(*parts)
+
+	def splitPath(self, path):
+		if haveFS:
+			return fs.path.split(path)
+		else:
+			return os.path.split(path)
+
+	def directoryName(self, path):
+		if haveFS:
+			return fs.path.dirname(path)
+		else:
+			return os.path.dirname(path)
+
+	def relativePath(self, path, start):
+		if haveFS:
+			return fs.relativefrom(path, start)
+		else:
+			return os.path.relpath(path, start)
+
+	# ---------
+	# Existence
+	# ---------
+
+	def exists(self, path):
+		return self._fsExists(path)
+
+	def isDirectory(self, path):
+		return self._fsIsDirectory(path)
+
+	def listDirectory(self, path, recurse=False):
+		return self._listDirectory(path, recurse=recurse, relativeTo=path)
+
+	def _listDirectory(self, path, recurse=False, relativeTo=None, depth=0, maxDepth=100):
+		sep = os.sep
+		if not relativeTo.endswith(sep):
+			relativeTo += sep
+		if depth > maxDepth:
+			raise UFOLibError("Maximum recusion depth reached.")
+		result = []
+		for fileName in self._fsListDirectory(path):
+			p = self.joinPath(path, fileName)
+			if self.isDirectory(p) and recurse:
+				result += self._listDirectory(p, recurse=True, relativeTo=relativeTo, depth=depth+1, maxDepth=maxDepth)
+			else:
+				p = p[len(relativeTo):]
+				if sep != "/":
+					# replace '\\' with '/'
+					p = p.replace(sep, "/")
+				result.append(p)
+		return result
+
+	def makeDirectory(self, path):
+		if not self.exists(path):
+			self._fsMakeDirectory(path)
+
+	# -----------
+	# File Opener
+	# -----------
+
+	def open(self, path, mode="r", encoding=None):
+		"""
+		Returns a file (or file-like) object for the
+		file at the given path. The path must be relative
+		to the UFO path. Returns None if the file does
+		not exist and the mode is "r" or "rb. An encoding
+		may be passed if needed.
+
+		Note: The caller is responsible for closing the open file.
+		"""
+		if encoding:
+			if mode == "r":
+				mode = "rb"
+			elif mode == "w":
+				mode = "wb"
+		if mode in ("r", "rb") and not self.exists(path):
+			return None
+		if self.exists(path) and self.isDirectory(path):
+			raise UFOLibError("%s is a directory." % path)
+		if mode in ("w", "wb"):
+			self._buildDirectoryTree(path)
```

Comment:
why should the `open` method need to create the directory tree for a file opened for writing? I know that if the parent folders of the file don't exist, an IOError will be raised dowstream. But that's how it should be, somebody else should be responsible for explicitly creating the tree, instead of silently and always creating them when calling open().